tests: drivers: uart_async_api: Be more robust to early buf release

talih0 opened the pull request:
Commit eb44414af99a44447afa52bb81ac0767f2ff791e modified 
test_single_read for early rx buf release.
    
The commit assumed that tdata.last_rx_buf points to either &tdata.rx_buf[0]
if the driver releases the first buffer or at &rx_buf[5] if the first buffer
is not released. However, where tdata.last_rx_buf points to depends on the
timeout given to uart_rx_enable(), making the test flaky.
    
This commit modifies the test by keeping track how many have been received
in the first and second buffers. The function tdata_check_recv_buffer()
validates that the sent data matches the received bytes which may have been
split between first and second buffer.
    
This fixes the uart_async_api test on the xmc45_relax_kit.

anangl commented on tests/drivers/uart/uart_async_api/src/test_uart_async.c:
```diff
@@ -127,7 +127,7 @@ ZTEST_USER(uart_async_single_read, test_single_read)
 			  "Initial buffer check failed");
 
-	uart_rx_enable(uart_dev, rx_buf, 10, 50 * USEC_PER_MSEC);
+	uart_rx_enable(uart_dev, rx_buf, 5, 50 * USEC_PER_MSEC);
```
RX buffer here is intentionally larger than TX buffer because particular UART drivers may or may not release the RX buffer after the first transfer (this is the whole point of #57847 where commit eb44414af99a44447afa52bb81ac0767f2ff791e was introduced) and this test is supposed to check if both transfers are successful, regardless of whether the initial RX buffer is released after the first transfer.
For the same reason, the `memcmp` at the end needs to check `tdata.last_rx_buf`, not `tdata.rx_buf`, as some drivers may receive both transfers to the first RX buffer (that `rx_buf` on the stack) and some may use the second one (`tdata.rx_buf`).

talih0 replied:
> For the same reason, the `memcmp` at the end needs to check `tdata.last_rx_buf`, not `tdata.rx_buf`, as some drivers may receive both transfers to the first RX buffer (that `rx_buf` on the stack) and some may use the second one (`tdata.rx_buf`).

If the buffer is not released, then tdata.last_rx_buf does not have to point to &rx_buf[5]. It can point to somewhere in the range &rx_buf[5]  to &rx_buf[9]. It depends on the timeout in uart_rx_enable() call.

talih0 replied:
An alternative could be to add a sleep after the first uart_tx call.

anangl replied:
> If the buffer is not released, then tdata.last_rx_buf does not have to point to &rx_buf[5]. It can point to somewhere in the range &rx_buf[5] to &rx_buf[9]. It depends on the timeout in uart_rx_enable() call.

Right. Currently, the test assumes that there won't be gaps longer than 50 ms between consecutive transmitted bytes, but it actually cannot be guaranteed. So indeed, this `last_rx_buf` pointer is not the best idea for this second check and this could be improved.

> An alternative could be to add a sleep after the first uart_tx call.

I'm not sure I understand what you mean. The test already waits 100 ms before the second TX: https://github.com/zephyrproject-rtos/zephyr/blob/312b1f9888b5e0e27e3cc511131b60bb0f171948/tests/drivers/uart/uart_async_api/src/test_uart_async.c#L137-L138

talih0 replied:
Yes, you are right.

talih0 replied:
> Right. Currently, the test assumes that there won't be gaps longer than 50 ms between consecutive transmitted bytes, but it actually cannot be guaranteed, so indeed, this `last_rx_buf` pointer is not the best idea for that second transfer check and this could be improved.

And if the timeout occurs before all the 5 bytes are received in the second transfer, we'll generate two RX_RDY calls, which will make validating the test messier.

dcpleung commented:
@talih0 any updates?

talih0 commented:
> @talih0 any updates?

I'll send an updated patch tomorrow.

talih0 committed:
tests: drivers: uart_async_api: Be more robust to early buf release

Commit eb44414af99a44447afa52bb81ac0767f2ff791e modified
test_single_read for early rx buf release.

The commit assumed that tdata.last_rx_buf points to either &tdata.rx_buf[0]
if the driver releases the first buffer or at &rx_buf[5] if the first
buffer is not released. However, where tdata.last_rx_buf points to depends
on the timeout given to uart_rx_enable(), making the test flaky.

This commit modifies the test by keeping track how many bytes have been
received in the first and second buffers.
The function tdata_check_recv_buffer() validates that the sent data
matches the received bytes which may have been split between first and
second buffer.

This fixes the uart_async_api test on the xmc45_relax_kit.

Signed-off-by: Andriy Gelman <andriy.gelman@gmail.com>

talih0 commented:
@anangl @dcpleung pls have a look when you have a moment.

dcpleung commented on tests/drivers/uart/uart_async_api/src/test_uart_async.c:
```diff
@@ -79,14 +79,34 @@ static void uart_async_test_init(void)
+
+static bool test_single_read_is_buf_in_first_buffer(const uint8_t *buf)
+{
+	uint32_t addr_buf = (uint32_t)buf;
```
Need to use `uintptr_t` instead to support 64-bit addressing.

talih0 replied:
thanks, I updated the PR.

dcpleung reviewed (changes requested):
Just a nit about the address type.

talih0 commented:
ping @anangl @nordic-krch

talih0 commented:
ping @anangl @nordic-krch

talih0 commented on tests/drivers/uart/uart_async_api/src/test_uart_async.c:
```diff
@@ -79,14 +79,34 @@ static void uart_async_test_init(void)
+
+static bool test_single_read_is_buf_in_first_buffer(const uint8_t *buf)
+{
+	uintptr_t addr_buf = (uintptr_t)buf;
+	uintptr_t first_buffer_start = (uintptr_t)tdata.rx_first_buffer;
+	uintptr_t first_buffer_end = first_buffer_start + sizeof(tdata.rx_first_buffer) - 1;
```
It's undefined behavior to compare pointers that do not belong to the same array (exceptions are NULL and one past the end of an array).

anangl replied:
I must admit that I was not aware of this and today I've learnt something new. Thanks!

But still, is using of this function actually better than just `if ((uintptr_t)evt->data.rx.buf < (uintptr_t)tdata.rx_second_buffer)`?

talih0 replied:
thanks, @anangl, I'll have another look tomorrow.

talih0 replied:
@anangl, I removed the extra function as you suggested. thanks.

talih0 commented on tests/drivers/uart/uart_async_api/src/test_uart_async.c:
```diff
@@ -196,11 +249,13 @@ ZTEST_USER(uart_async_multi_rx, test_multiple_rx_enable)
+	uint32_t rx_buf_size = sizeof(tx_buf);
 	int ret;
 
+	zassert_equal(rx_buf_size <= sizeof(tdata.rx_first_buffer), 1, "Invalid buf size");
```
thanks, will update.
